[Nexus] feature to use JSON file about usable presets and allow to change by user

AlwinEsch opened the pull request:
This is a nice feature to allow an user to create own preset file and his own textures(images about.
As I like the request https://github.com/xbmc/visualization.shadertoy/pull/41 and there not updated, thought to make byself :smile:.

Here how the file can be used outside of addon (also used for test):
```*.json
{ "presets":[
[ "Waves Remix",  "./WavesRemix.glsl", "audio", "", "", "" ],
[ "The Disco Tunnel", "./TheDiscoTunnel.frag.glsl",  "./tex02.png", "./tex10.png", "audio", "" ]
]}
```

Here addon his own where his previous parts are moved from code to there:
```*.json
{ "presets":[
[ "2D LED Spectrum by un1versal",             "2Dspectrum.frag.glsl",             "audio",     "",          "",      "" ],
[ "Audio Eclipse by airtight",                "audioeclipse.frag.glsl",           "audio",     "",          "",      "" ],
[ "Audio Reaktive by choard1895",             "audioreaktive.frag.glsl",          "audio",     "",          "",      "" ],
[ "AudioVisual by Passion",                   "audiovisual.frag.glsl",            "audio",     "",          "",      "" ],
[ "Beating Circles by Phoenix72",             "beatingcircles.frag.glsl",         "audio",     "",          "",      "" ],
[ "BPM by iq",                                "bpm.frag.glsl",                    "audio",     "",          "",      "" ],
[ "Circle Wave by TekF",                      "circlewave.frag.glsl",             "audio",     "",          "",      "" ],
[ "Circuits by Kali",                         "circuits.frag.glsl",               "audio",     "tex07.png", "",      "", "gl_only" ],
[ "Colored Bars by novalis",                  "coloredbars.frag.glsl",            "audio",     "",          "",      "", "gl_only" ],
[ "Cubescape by iq",                          "cubescape.frag.glsl",              "audio",     "tex05.png", "",      "", "gl_only" ],
[ "Dancing Metalights by Danguafare",         "dancingmetalights.frag.glsl",      "audio",     "",          "",      "" ],
[ "The Disco Tunnel by poljere",              "discotunnel.frag.glsl",            "tex02.png", "tex15.png", "audio", "" ],
[ "Electric pulse by un1versal",              "electricpulse.frag.glsl",          "audio",     "",          "",      "" ],
[ "Fractal Land by Kali",                     "fractalland.frag.glsl",            "tex02.png", "tex15.png", "audio", "", "gl_only" ],
[ "Gameboy by iq",                            "gameboy.frag.glsl",                "audio",     "",          "",      "" ],
[ "Input Sound by iq",                        "input.frag.glsl",                  "audio",     "",          "",      "" ],
[ "I/O by movAX13h",                          "io.frag.glsl",                     "audio",     "",          "",      "", "gl_only" ],
[ "Kaleidoscope Visualizer by Qqwy",          "kaleidoscopevisualizer.frag.glsl", "audio",     "tex14.png", "",      "" ],
[ "LED spectrum by simesgreen",               "ledspectrum.frag.glsl",            "audio",     "",          "",      "" ],
[ "Polar Beats by sauj123",                   "polarbeats.frag.glsl",             "audio",     "",          "",      "" ],
[ "Simplicity Galaxy by CBS",                 "simplicitygalaxy.frag.glsl",       "audio",     "",          "",      "" ],
[ "Sound Flower by iq",                       "soundflower.frag.glsl",            "audio",     "",          "",      "" ],
[ "Sound sinus wave by Eitraz",               "soundsinuswave.frag.glsl",         "audio",     "",          "",      "" ],
[ "Spectrometer by jaba",                     "spectrometer.frag.glsl",           "audio",     "",          "",      "" ],
[ "symmetrical sound visualiser by thelinked","symmetricalsound.frag.glsl",       "audio",     "",          "",      "" ],
[ "Twisted Rings by poljere",                 "twistedrings.frag.glsl",           "audio",     "",          "",      "" ],
[ "Undulant Spectre by mafik",                "undulantspectre.frag.glsl",        "audio",     "",          "",      "" ],
[ "Demo - Volumetric Lines by iq",            "volumetriclines.frag.glsl",        "audio",     "",          "",      "", "gl_only" ],
[ "Waves Remix by ADOB",                      "wavesremix.frag.glsl",             "audio",     "",          "",      "" ]
]}
```

Also added a additional python script to allow show of help text, so that a user not need to search somewhere how works.
Think a english speaking guy can be look to make it better in grammar :sweat_smile: .

Here windows how them now looks:
![Bildschirmfoto vom 2021-09-06 00-42-40](https://user-images.githubusercontent.com/6879739/132143589-bec500f3-97a1-4bb5-8d27-bb1f159db104.png)
![Bildschirmfoto vom 2021-09-06 00-35-29](https://user-images.githubusercontent.com/6879739/132143598-e733bcc9-4248-4b2b-8b93-87b9093fcd17.png)
![Bildschirmfoto vom 2021-09-06 00-46-42](https://user-images.githubusercontent.com/6879739/132143650-0eb22e6b-02fc-4c5d-85ca-38a4ac45d9ce.png)

And can be opened with:
![Bildschirmfoto vom 2021-09-06 00-34-19](https://user-images.githubusercontent.com/6879739/132143608-f8dc6eea-db03-43e8-86d9-8440588f8786.png)

Thought to make shortly a complete normal programming, instead of the update round to change all addons to Nexus :smirk::rofl: , where becomes now continued.

AlwinEsch commented:
Has forgot, in case of a wrong JSON, missing files and problems becomes a notification shown (and detailed on log) to know why not work.
![Bildschirmfoto von 2021-09-06 01-33-48](https://user-images.githubusercontent.com/6879739/132144434-8911ded1-30f0-4e81-9ccc-553a273caba6.png)

AlwinEsch committed:
add jsoncpp to depends

chewitt commented on visualization.shadertoy/changelog.txt:
```diff
@@ -1,3 +1,9 @@
+  - There one as standard with 'presets.json' on addon
+  - Further can user set within settings to a own file
+  - Also to help about is within settings a button added to open text window with description about
+
```
- Add new custom shader list feature:
  - Move existing shader list from add-on code to the default 'presets.json' file
  - Allow users to configure a path to their own custom preset file
  - Add help button with instructions on preset file JSON formatting

AlwinEsch replied:
Sounds much better, has changed too.

chewitt commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -123,3 +123,39 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30032"
+msgid "Path to own preset list"
```
Path to custom preset list

chewitt commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -123,3 +123,39 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30031"
+msgid "Incorrect reading of the given JSON file, check whether it is correct and available."
```
Cannot read the JSON file. Please check the formatting is correct and the file is available.

chewitt commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -123,3 +123,39 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30033"
+msgid "Within here can be an own path defined to a preset list within JSON format. If you leave empty, addon use his own preset file."
```
Set the path to your custom JSON formatted preset file. If no file path is configured the add-on uses the default preset file.

chewitt commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -123,3 +123,39 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30034"
+msgid "Define a single shader"
```
Configure a single shader

chewitt commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -123,3 +123,39 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30036"
+msgid "This open a text viewer where gives help and how a preset file in JSON format need be defined."
```
Open to read information on how-to build the JSON formatted preset file.

ProfYaffle replied:
"how to", not "how-to" (verb, not compound noun).

AlwinEsch replied:
Thanks a lot @chewitt and @ProfYaffle to look over it, has changed on all places to your ones.

ProfYaffle commented on visualization.shadertoy/resources/presets.json.md:
```diff
@@ -0,0 +1,39 @@
+# User definable shader list within JSON file
```
This whole section has been reviewed by @howie-f , so refer to this commit for changes:

https://github.com/howie-f/visualization.shadertoy/commit/dfc7c92b1d8eb784cee00837e000db985797adf2

ProfYaffle commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -123,3 +123,39 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30038"
+msgid "[B][I]User definable shader list within JSON file[/I][/B][CR][CR]A e.g. [B]presets.json[/B] can be used to set own shaders and related parts by self.[CR]Within addon settings can be this set to an own path where can be somewhere else.[CR][CR]It should be:[CR][I]  1.[/I] Name of preset, where could be text or number where set within [I]strings.po[/I][CR][I]  2.[/I] glsl shader to use, if pa ...
```
I think this entire string repeats the text in presets.json.md, below. As such, all edits to one should simply copy over once finalised.

AlwinEsch committed:
increase version to 20.1.0

AlwinEsch commented on visualization.shadertoy/resources/presets.json.md:
```diff
@@ -0,0 +1,42 @@
+/* Select texture image by file name, use default `resources/` dir */
+[ "Fractal Land by Kali",         "fractalland.frag.glsl",      "tex02.png", "tex15.png", "audio", "" ]
+]}
+~~~
```
Have it updated. For here and also strings.po are texts complete equal, only about the format are there differences and for this file much better human readable :sweat_smile: .

Only has leaved the additional line break from strings.po away, as there then nice visible in single page view:
![Bildschirmfoto von 2021-09-10 22-12-54](https://user-images.githubusercontent.com/6879739/132912363-a6a6a42f-5813-4b2b-980a-ff1a8bcb651d.png).

Further has leaved the channel numbers by begin of 0, also not likes to have 0 as first, only use ShaderToy it the same, e.g. https://www.shadertoy.com/view/4ljGD1 and also on other settings already with begin as 0.

Thanks a lot @howie-f :+1:

eigendude commented on src/PresetsLoader.cpp:
```diff
@@ -0,0 +1,175 @@
+        const auto& entry = root["presets"][index];
+        Preset preset;
+
+        // Check optional value about support limitation added
```
This comment is worded a bit creatively, can you explain what it means? I'll help get it understandable for everyone

EDIT: Eigendude is my work account, it's garbear

AlwinEsch replied:
Yes sounds really bad. Maybe here a bit better: `Check the optional JSON value at the end of the preset, which sets possible limits, e.g. support only to GL ("gl_only").` 

Thought to make there below the code about checks and to hold the description text a bit more general, for case another checks comes to it in future.

garbear replied:
OK how about

```c++
// Check the optional JSON value at the end of the preset, which sets possible limits, e.g. supporting only GL ("gl_only")

AlwinEsch replied:
Thanks, has changed to it.

eigendude commented on src/PresetsLoader.cpp:
```diff
@@ -0,0 +1,175 @@
+  if (ptr)
+    delete[] ptr;
+
+  /* Inform user that's maybe his own preset list is wrong. */
```
```c++
// Inform user that maybe their own preset list is wrong
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "iChannel 3"
 msgstr ""
 
+#. Boolean settings value to enable and allow selection about own ".glsl" file and related textures
```
```
#. Boolean settings value to enable and allow selection of a ".glsl" file and related textures
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "Use your own shadertoy. Attention: Some shaders are not supported by some GPU's in performance!"
 msgstr ""
 
+#. To open file selection dialog, about *.glsl file, if setting "Choose one yourself"(30010) enabled.
```
```
#. To open the file selection dialog about *.glsl files, if the setting "Choose one yourself"(30010) above is enabled.
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "Shader as *.glsl"
 msgstr ""
 
+#. Boolean value to use audio on channel, if setting "Choose one yourself"(30010) enabled.
```
```
#. Boolean value to use channel 0 as audio, if the setting "Choose one yourself"(30010) above is enabled.
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "Use audio stream data on channel 0"
 msgstr ""
 
+#. Texture image to use if setting "Choose one yourself"(30010) enabled.
```
```
#. Texture image to use, if the setting "Choose one yourself"(30010) above is enabled.
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "Channel 0 as *.png"
 msgstr ""
 
+#. Boolean value to use audio on channel, if setting "Choose one yourself"(30010) enabled.
```
```
#. Boolean value to use channel 1 as audio, if the setting "Choose one yourself"(30010) above is enabled.
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "Use audio stream data on channel 1"
 msgstr ""
 
+#. Texture image to use if setting "Choose one yourself"(30010) enabled.
```
```
#. Texture image to use, if the setting "Choose one yourself"(30010) above is enabled.
```

and below

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
 msgid "Channel 1 as *.png"
 msgstr ""
 
+#. Boolean value to use audio on channel, if setting "Choose one yourself"(30010) enabled.
```
Same as above with "channel 2" instead, and below

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+msgid "Shadertoy Visualization"
+msgstr ""
+
+#. Notification message shown on screen by wrong preset list
```
```
#. Notification message shown on the screen when an incorrect preset list is given
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+#. Notification message shown on screen by wrong preset list
+#: visualization.shadertoy/src/PresetsLoader.cpp
+msgctxt "#30031"
+msgid "Cannot read the JSON file. Please check the formatting is correct and the file is available."
```
```
msgid "Cannot read the JSON file. Please check that the formatting is correct and the file is available."

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+#. Help text related to setting string with id 30032
+#: visualization.shadertoy/resources/settings.xml
+msgctxt "#30033"
+msgid "Set the path to your custom JSON formatted preset file. If no file path is configured the add-on uses the default preset file."
```
```
msgid "Set the path to your custom JSON-formatted preset file. If no path is configured, the add-on will use the default preset file."
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+msgid "Configure a single shader"
+msgstr ""
+
+#. Settings value to open dialog for show of JSON format help on preset file
```
```
#. Settings value to open a dialog for showing help about the preset file
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+#. Help text related to setting string with id 30035
+#: visualization.shadertoy/resources/settings.xml
+msgctxt "#30036"
+msgid "Open to read information on how to build the JSON formatted preset file."
```
```
msgid "Open to read about creating the JSON-formatted preset file."
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+msgid "Open to read information on how to build the JSON formatted preset file."
+msgstr ""
+
+#. Header text on help dialog opened within settings, related help text on 30038
```
```
#. Header text on the help dialog within settings, related to the help text of string ID 30038
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+#. Header text on help dialog opened within settings, related help text on 30038
+#: visualization.shadertoy/resources/settings.xml
+msgctxt "#30037"
+msgid "Help about JSON file to define own shaders"
```
```
msgid "Help about defining your own shaders in the JSON file"
```

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+#. only in Kodi's format style and colored code example!
+#: visualization.shadertoy/resources/settings.xml
+msgctxt "#30038"
+msgid "[B][I]JSON file for user defined shader lists[/I][/B][CR][CR]The file [B]presets.json[/B] is used to define custom shaders.[CR]A custom path to the shader files (GLSL) can be declared in the add-on settings.[CR][CR]Contents of [B]presets.json[/B] should be:[CR][I]  1.[/I] Name of preset. Either hardcoded string or defined in [I]resources/language/.../strings.po[/I][CR][I]  2.[/I] GLSL shad ...
```
Is such a long string with so many tags possible to translate?

AlwinEsch replied:
Not sure, thing by this maybe some says no, as really hard to understand what's there inside.
Only not found a way to make easy readable by user and also nice shown on text window on Kodi. Specially this color set on code example, makes it much more easy and understandable for Users on view by Window.

@gade01 how does it works with strings.po and added linebreaks, by scripts and Weblate?

Kodi already support as e.g.:
```po
msgctxt "#30038"
msgid "[B][I]JSON file for user defined shader lists[/I][/B]\n"
"\n"
"The file [B]presets.json[/B] is used to define custom shaders.\n"
"A custom path to the shader files (GLSL) can be declared in the add-on settings.[CR][CR]Contents of [B]presets.json[/B] should be:\n"
"[I]  1.[/I] Name of preset. Either hardcoded string or defined in [I]resources/language/.../strings.po[/I]\n"
"[I]  2.[/I] GLSL shader to use. if no custom path was declared, the default [B]resources/[/B] will be used. Set custom path to './' to use shaders from the same path as your [B]presets.json[/B] file.\n"
"[I]  3.[/I] Channel value 0. Defines the texture to use (e.g. PNG file). Can also be set to \"audio\" to pass related stream data within texture. The PNG and GLSL files should be located in the same directory.\n"
"[I]  4.[/I] Channel value 1 definition.\n"
"[I]  5.[/I] Channel value 2 definition.\n"
"[I]  6.[/I] Channel value 3 definition.\n"
"[I]  7.[/I] [I][Optional][/I] append string [B]\"gl_only\"[/B] to make this preset only available to OpenGL.\n"
"\n"
"[B]Example:[/B]\n"
"[COLOR white][COLOR green]{[/COLOR] [COLOR blue]\"presets\"[/COLOR]:[\n"
"[COLOR gray]/* Default way used by the addon itself */[/COLOR]\n"
"[ [COLOR red]\"The Disco Tunnel by poljere\"[/COLOR],  [COLOR red]\"discotunnel.frag.glsl\"[/COLOR], [COLOR red]\"tex02.png\"[/COLOR], [COLOR red]\"tex15.png\"[/COLOR], [COLOR red]\"audio\"[/COLOR], [COLOR red]\"\"[/COLOR] ],\n"
"[COLOR gray]/* Description defined in strings.po */[/COLOR]\n"
"[ 30100, [COLOR red]\"audioeclipse.frag.glsl\"[/COLOR], [COLOR red]\"audio\"[/COLOR], [COLOR red]\"\"[/COLOR], [COLOR red]\"\"[/COLOR], [COLOR red]\"\"[/COLOR] ],\n"
"[COLOR gray]/* Define custom file paths */[/COLOR]\n"
"[ [COLOR red]\"My own one\"[/COLOR], [COLOR red]\"/path/to/glsl/my.frag.glsl\"[/COLOR], [COLOR red]\"audio\"[/COLOR], [COLOR red]\"/path/to/pngs/my_image.png\"[/COLOR], [COLOR red]\"\"[/COLOR], [COLOR red]\"\"[/COLOR] ],\n"
"[COLOR gray]/* Preset only available to OpenGL (see 7)  */[/COLOR]\n"
"[ [COLOR red]\"Polar Beats by sauj123\"[/COLOR], [COLOR red]\"polarbeats.frag.glsl\"[/COLOR], [COLOR red]\"audio\"[/COLOR], [COLOR red]\"\"[/COLOR], [COLOR red]\"\"[/COLOR], [COLOR red]\"\"[/COLOR], [COLOR red]\"gl_only\"[/COLOR] ],\n"
"[COLOR gray]/* GLSL and PNG files are in the same folder as presets.json */[/COLOR]\n"
"[ [COLOR red]\"Another one\"[/COLOR], [COLOR red]\"./another.frag.glsl\"[/COLOR], [COLOR red]\"./my_01.png\"[/COLOR], [COLOR red]\"./my_02.png\"[/COLOR], [COLOR red]\"audio\"[/COLOR], [COLOR red]\"\"[/COLOR] ],\n"
"[COLOR gray]/* Select texture image by file name, use default \"resources/\" dir */[/COLOR]\n"
"[ [COLOR red]\"Fractal Land by Kali\"[/COLOR], [COLOR red]\"fractalland.frag.glsl\"[/COLOR], [COLOR red]\"tex02.png\"[/COLOR], [COLOR red]\"tex15.png\"[/COLOR], [COLOR red]\"audio\"[/COLOR], [COLOR red]\"\"[/COLOR] ]\n"
"][COLOR green]}[/COLOR][/COLOR]"
msgstr ""
```

But seen that you changed everywhere to a single line and I thought other ways not possible.

Also a part where I hope to become "Markdown" support added to Kodi, for cases like there (where planned from me since years :cry: ).

eigendude commented on visualization.shadertoy/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -9,117 +9,228 @@ msgstr ""
+msgid "Help about JSON file to define own shaders"
+msgstr ""
+
+#. Help text about preset file in JSON format to show on text window.
```
```
#. Help text about showing the JSON-formatted preset file in a text window.
```

(I think)

eigendude commented on visualization.shadertoy/resources/presets.json.md:
```diff
@@ -0,0 +1,42 @@
+
+Contents of `presets.json` should be:
+1. Name of preset. Either hardcoded string or defined in `resources/language/.../strings.po`
+2. GLSL shader to use. if no custom path was declared, the default `resources/` will be used.
```
Capital `if`

eigendude commented:
I think I caught most of the grammar improvements

AlwinEsch committed:
allow JSON preset file selection via settings

AlwinEsch committed:
add feature to open text window via settings to show help for JSON

Further is a presets.json.md added to show related text on github too.

AlwinEsch committed:
allow load of available presets via a JSON file

This thought to make available types more easy selectable for a user.

AlwinEsch commented:
Then think about bringing it in and creating a backport for it.

AlwinEsch committed:
add help text to other unchanged strings within strings.po

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translated using Weblate (Chinese (China) (zh_cn))

Currently translated at 3.7% (1 of 27 strings)

Translated using Weblate (Korean (ko_kr))

Currently translated at 3.7% (1 of 27 strings)

Co-authored-by: Hosted Weblate <hosted@weblate.org>
Co-authored-by: Minho Park <parkmino@gmail.com>
Co-authored-by: taxigps <taxigps@sina.com>
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-look-and-feel/visualization-shadertoy/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-look-and-feel/visualization-shadertoy/ko_kr/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-look-and-feel/visualization-shadertoy/zh_cn/
Translation: Kodi add-ons: look and feel/visualization.shadertoy
